packages/alfa-math/src/matrix.ts

```diff
@@ -1,63 +1,175 @@
-  ): Matrix<M, N> {
-    return matrix.map((vector) => Vector.clone(vector)) as Matrix<M, N>;
+  export function clone(m: Matrix): Matrix {
+    return m.map((r) => r.slice(0));
   }
 
-  export function multiply<M extends number, N extends number>(
-    matrix: Matrix<M, N>,
-    vector: Vector<N>
-  ): Vector<N>;
+  export function equals(m: Matrix, n: Matrix, e?: number): boolean {
+    return (
+      m.length === n.length &&
+      m.every(
+        (r, i) =>
+          r.length === n[i].length &&
+          r.every((v, j) => Real.equals(v, n[i][j], e))
+      )
+    );
+  }
+
+  export function identity(n: number): Matrix {
+    return new Array<Array<number>>(n).fill([]).map((_, i) => {
+      const r = new Array<number>(n).fill(0);
+      r[i] = 1;
+      return r;
+    });
+  }
+
+  export function size(m: Matrix): [number, number] {
+    return [rows(m), columns(m)];
+  }
+
+  export function rows(m: Matrix): number {
+    return m.length;
+  }
+
+  export function columns(m: Matrix): number {
+    return m.length === 0 ? 0 : m[0].length;
+  }
+
+  export function row(m: Matrix, i: number): Array<number> {
+    return m[i];
+  }
+
+  export function column(m: Matrix, i: number): Array<number> {
+    return m.map((r) => r[i]);
+  }
+
+  export function add(m: Matrix, n: Matrix): Matrix {
+    return m.map((r, i) => r.map((v, j) => v + n?.[i]?.[j]));
+  }
+
+  export function subtract(m: Matrix, n: Matrix): Matrix {
+    return m.map((r, i) => r.map((v, j) => v - n?.[i]?.[j]));
+  }
+
+  /**
+   * @see https://en.wikipedia.org/wiki/Matrix_multiplication
+   */
+  export function multiply(m: Matrix, n: number | Matrix): Matrix {
+    return typeof n === "number"
+      ? m.map((r) => r.map((v) => v * n))
+      : m.map((r, i) =>
+          n[0].map((_, j) =>
+            r.reduce((s, _, k) => s + m[i][k] * n?.[k]?.[j], 0)
+          )
+        );
+  }
+
+  /**
+   * Compute the transpose of a matrix.
+   *
+   * @see https://en.wikipedia.org/wiki/Transpose
+   */
+  export function transpose(m: Matrix): Matrix {
+    return m.length === 0 ? m : m[0].map((_, i) => m.map((row) => row[i]));
+  }
 
-  export function multiply<
-    N extends number,
-    M extends number,
-    P extends number
-  >(a: Matrix<N, M>, b: Matrix<M, P>): Matrix<N, P>;
+  /**
+   * Compute the determinant of a non-empty, square matrix.
+   *
+   * @see https://en.wikipedia.org/wiki/Determinant
+   * @see https://en.wikipedia.org/wiki/Laplace_expansion
+   *
+   * @remarks
+   * This function uses Laplace expansion for computing the determinant which
+   * has a time complexity of O(n!) and is therefore not practical for large
+   * matrices.
+   */
+  export function determinant(m: Matrix): number {
+    switch (m.length) {
+      // The determinant of a 1x1 matrix is the single value of the matrix.
+      case 1:
+        return m[0][0];
+
+      // The determinant of a 2x2 matrix is fairly short to write out and so
+      // this is done as an optimization.
```

Comment:
I don't remember precisely (might be too big to write), but should we have a special case for 4×4 matrices given that it is our most common use case?

Reply:
If they aren't too cumbersome to write out we can absolutely include cases for 3x3 and 4x4 matrices 👍